# 4. 구현

파이썬 기본 문법에 능해야 풀 때 유리. ex. itertools

## 예제 4-1. 상하좌우

![예제4-1_1](./image/예제4-1_1.png)
![예제4-1_2](./image/예제4-1_2.png)

In [2]:
n = int(input())
root = list(input().split())

In [3]:
root

['R', 'R', 'R', 'U', 'D', 'D']

In [6]:
row = 1 # 현재 위치 1,1
col = 1
for i in root:
    if i == 'R': # R이면 오른쪽으로 한칸 이동. 즉, col +1. 만약 현재 위치가 n이라면 이동 x.
        if col != n:
            col += 1
    elif i == 'L': # L이면 왼쪽으로 한칸 이동. 즉, col -1. 만약 현재 위치가 1이라면 이동 x.
        if col != 1:
            col -= 1
    elif i == 'U': # U이면 위로 한칸 이동. 즉, row -1. 만약 현재 위치가 1이라면 이동 x.
        if row != 1:
            row -= 1
    else: # U이면 아래로 한칸 이동. 즉, row +1. 만약 현재 위치가 n이라면 이동 x.
        if row != n:
            row += 1
print(row, col)

3 4


In [ ]:
# 책 풀이
n = int(input())
x, y = 1, 1
plans = input().split()

# L, R, U, D에 따른 이동방향
dx = [0,0,-1,1]
dy = [-1,1,0,0]
move_types = ['L', 'R', 'U', 'D']

# 이동계획을 하나씩 확인
for plan in plans:
    # 이동 후 좌표 구하기
    for i in range(len(move_types)):
        if plan == move_types[i]:
            nx = x + dx[i]
            ny = y + dy[i]
    # 공간을 벗어나는 경우 무시
    if nx < 1 or ny < 1 or nx > n or ny > n:
        continue
    # 이동 수행
    x, y = nx, ny

print(x,y)

## 예제 4-2. 시각
![예제4-2](./image/예제4-2.png)

In [8]:
n = int(input())

count = 0
for i in range(n+1):
    for j in range(60):
        for k in range(60):
            # 매시각 안에 '3'이 포함되어 있다면 카운트 증가
            if '3' in str(i) + str(j) + str(k):
                count += 1
print(count)

11475


## 실전문제
### 왕실의 나이트
![4-왕실의나이트](./image/4-왕실의나이트.png)
![4-왕실의나이트2](./image/4-왕실의나이트2.png)

In [4]:
input_data = input()
row = int(input_data[1])
column = int(ord(input_data[0])) - int(ord('a')) + 1 # a 의 순서와의 차이에 1을 더하면 현재 열 위치 얻을 수 있음

# 나이트의 이동 경로 
steps = [(-2,-1), (-1,-2), (1,-2), (2,-1), (2,1), (1,2), (-1,2), (-2,1)] # 총 8가지

# 나이트의 현재 위치가 주어지면 현재 위치에서 이동 경로를 더한 다음, 8 x 8 좌표 평면에 있는지 확인하면 됨.
cases = 0
for step in steps:
    loc_row = row + step[0]
    loc_col = column + step[1]
    if loc_row >= 1 and loc_row <=8 and loc_col >= 1 and loc_col <= 8:
        cases += 1
print(cases)

2


### 게임개발
![4-게임개발1](./image/4-게임개발1.png)
![4-게임개발2](./image/4-게임개발2.png)

In [5]:
n, m = map(int, input().split())
# 방문한 위치를 저장하기 위한 맵을 생성하여 0으로 초기화
d = [[0]*m for _ in range(n)] # n x m 맵 생성
# 현재 캐릭터의 x좌표,  y좌표, 방향 입력받기
x, y, direction = map(int, input().split())
d[x][y] = 1 # 현재 좌표 방문 처리

# 전체 맵 정보 입력받기
array = []
for i in range(n):
    array.append(list(map(int, input().split())))

# 북, 동, 남, 서 방향 정의
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

# 왼쪽으로 회전
def turn_left():
    global direction
    direction -= 1
    if direction == -1:
        direction = 3
    
# 시뮬레이션 시작
count = 1
turn_time = 0
while True:
    # 왼쪽으로 회전
    turn_left()
    nx = x + dx[direction]
    ny = y + dy[direction]
    # 회전한 이후 정면에 가보지 않은 칸이 존재하는 경우 이동
    if d[nx][ny] == 0 and array[nx][ny] == 0: # 가보지 않았고 육지인 경우
        d[nx][ny] = 1
        x = nx
        y = ny
        count += 1
        turn_time = 0
        continue
    # 회전한 이후 정면에 가보지 않은 칸이 없거나 바다인 경우
    else:
        turn_time += 1
    # 네 방향 모두 갈 수 없는 경우 뒤로 한 칸 이동
    if turn_time == 4:
        nx = x - dx[direction]
        ny = y - dy[direction]
        # 뒤로 갈 수 있다면 이동
        if array[nx][ny] == 0:
            x = nx
            y = ny
        # 뒤가 바다로 막혀 있는 경우
        else:
            break
        turn_time = 0
print(count)

3


# 백준

## 셀프넘버
- 문제\
셀프 넘버는 1949년 인도 수학자 D.R. Kaprekar가 이름 붙였다. 양의 정수 n에 대해서 d(n)을 n과 n의 각 자리수를 더하는 함수라고 정의하자. 예를 들어, d(75) = 75+7+5 = 87이다.
양의 정수 n이 주어졌을 때, 이 수를 시작해서 n, d(n), d(d(n)), d(d(d(n))), ...과 같은 무한 수열을 만들 수 있다. 
예를 들어, 33으로 시작한다면 다음 수는 33 + 3 + 3 = 39이고, 그 다음 수는 39 + 3 + 9 = 51, 다음 수는 51 + 5 + 1 = 57이다. 이런식으로 다음과 같은 수열을 만들 수 있다.\
33, 39, 51, 57, 69, 84, 96, 111, 114, 120, 123, 129, 141, ...\
n을 d(n)의 생성자라고 한다. 위의 수열에서 33은 39의 생성자이고, 39는 51의 생성자, 51은 57의 생성자이다. 생성자가 한 개보다 많은 경우도 있다. 예를 들어, 101은 생성자가 2개(91과 100) 있다. 
생성자가 없는 숫자를 셀프 넘버라고 한다. 100보다 작은 셀프 넘버는 총 13개가 있다. 1, 3, 5, 7, 9, 20, 31, 42, 53, 64, 75, 86, 97
10000보다 작거나 같은 셀프 넘버를 한 줄에 하나씩 출력하는 프로그램을 작성하시오.

- 입력\
입력은 없다.

- 출력\
10,000보다 작거나 같은 셀프 넘버를 한 줄에 하나씩 증가하는 순서로 출력한다.

In [17]:
self_num_lst = [0]*10036
for i in range(10000): # 생성자를 0부터 9999까지 -> 생성자가 9999일 때, 최대 셀프넘버 10035
    num = i+1 # 생성자
    sevaral_num = list(str(num)) # 생성자를 각 자릿수별로 나누어 리스트로 저장
    self_num = num
    for j in sevaral_num:
        self_num += int(j) # not셀프넘버 생성
    self_num_lst[self_num] += 1 # 셀프넘버에 해당하는 인덱스에 +1
# 리스트에서 값이 0인 인덱스가 셀프넘버가 아닌 숫자
for i in range(len(self_num_lst)):
    if i != 0 and self_num_lst[i] == 0 and i < 10000:
        print(i)

1
3
5
7
9
20
31
42
53
64
75
86
97
108
110
121
132
143
154
165
176
187
198
209
211
222
233
244
255
266
277
288
299
310
312
323
334
345
356
367
378
389
400
411
413
424
435
446
457
468
479
490
501
512
514
525
536
547
558
569
580
591
602
613
615
626
637
648
659
670
681
692
703
714
716
727
738
749
760
771
782
793
804
815
817
828
839
850
861
872
883
894
905
916
918
929
940
951
962
973
984
995
1006
1021
1032
1043
1054
1065
1076
1087
1098
1109
1111
1122
1133
1144
1155
1166
1177
1188
1199
1210
1212
1223
1234
1245
1256
1267
1278
1289
1300
1311
1313
1324
1335
1346
1357
1368
1379
1390
1401
1412
1414
1425
1436
1447
1458
1469
1480
1491
1502
1513
1515
1526
1537
1548
1559
1570
1581
1592
1603
1614
1616
1627
1638
1649
1660
1671
1682
1693
1704
1715
1717
1728
1739
1750
1761
1772
1783
1794
1805
1816
1818
1829
1840
1851
1862
1873
1884
1895
1906
1917
1919
1930
1941
1952
1963
1974
1985
1996
2007
2022
2033
2044
2055
2066
2077
2088
2099
2110
2112
2123
2134
2145
2156
2167
2178
2189
2200
2211
2213
2224
2235
2246


## 덩치
- 문제\
우리는 사람의 덩치를 키와 몸무게, 이 두 개의 값으로 표현하여 그 등수를 매겨보려고 한다. 어떤 사람의 몸무게가 x kg이고 키가 y cm라면 이 사람의 덩치는 (x, y)로 표시된다. 두 사람 A 와 B의 덩치가 각각 (x, y), (p, q)라고 할 때 x > p 그리고 y > q 이라면 우리는 A의 덩치가 B의 덩치보다 "더 크다"고 말한다. 예를 들어 어떤 A, B 두 사람의 덩치가 각각 (56, 177), (45, 165) 라고 한다면 A의 덩치가 B보다 큰 셈이 된다. 그런데 서로 다른 덩치끼리 크기를 정할 수 없는 경우도 있다. 예를 들어 두 사람 C와 D의 덩치가 각각 (45, 181), (55, 173)이라면 몸무게는 D가 C보다 더 무겁고, 키는 C가 더 크므로, "덩치"로만 볼 때 C와 D는 누구도 상대방보다 더 크다고 말할 수 없다.\
N명의 집단에서 각 사람의 덩치 등수는 자신보다 더 "큰 덩치"의 사람의 수로 정해진다. 만일 자신보다 더 큰 덩치의 사람이 k명이라면 그 사람의 덩치 등수는 k+1이 된다. 이렇게 등수를 결정하면 같은 덩치 등수를 가진 사람은 여러 명도 가능하다. 아래는 5명으로 이루어진 집단에서 각 사람의 덩치와 그 등수가 표시된 표이다.\
이름	(몸무게, 키)	덩치 등수\
A	(55, 185)	2\
B	(58, 183)	2\
C	(88, 186)	1\
D	(60, 175)	2\
E	(46, 155)	5\
위 표에서 C보다 더 큰 덩치의 사람이 없으므로 C는 1등이 된다. 그리고 A, B, D 각각의 덩치보다 큰 사람은 C뿐이므로 이들은 모두 2등이 된다. 그리고 E보다 큰 덩치는 A, B, C, D 이렇게 4명이므로 E의 덩치는 5등이 된다. 위 경우에 3등과 4등은 존재하지 않는다. 여러분은 학생 N명의 몸무게와 키가 담긴 입력을 읽어서 각 사람의 덩치 등수를 계산하여 출력해야 한다.

- 입력\
첫 줄에는 전체 사람의 수 N이 주어진다. 그리고 이어지는 N개의 줄에는 각 사람의 몸무게와 키를 나타내는 양의 정수 x와 y가 하나의 공백을 두고 각각 나타난다.

- 출력\
여러분은 입력에 나열된 사람의 덩치 등수를 구해서 그 순서대로 첫 줄에 출력해야 한다. 단, 각 덩치 등수는 공백문자로 분리되어야 한다.

In [18]:
n = int(input())
lst = [list(map(int, input().split())) for _ in range(n)]

In [27]:
rank = [0]*n
for i in range(n):
    weight = lst[i][0] 
    height = lst[i][1]
    for j in range(n): # 몸무게와 키가 모두 더 크면 rank +1
        if lst[j][0] > weight and lst[j][1] > height:
            rank[i] += 1
for i in rank:
    print(i+1, end=' ')

2 2 1 2 5 

## 덱
- 문제\
정수를 저장하는 덱(Deque)를 구현한 다음, 입력으로 주어지는 명령을 처리하는 프로그램을 작성하시오.
명령은 총 여덟 가지이다.\
push_front X: 정수 X를 덱의 앞에 넣는다.\
push_back X: 정수 X를 덱의 뒤에 넣는다.\
pop_front: 덱의 가장 앞에 있는 수를 빼고, 그 수를 출력한다. 만약, 덱에 들어있는 정수가 없는 경우에는 -1을 출력한다.\
pop_back: 덱의 가장 뒤에 있는 수를 빼고, 그 수를 출력한다. 만약, 덱에 들어있는 정수가 없는 경우에는 -1을 출력한다.\
size: 덱에 들어있는 정수의 개수를 출력한다.\
empty: 덱이 비어있으면 1을, 아니면 0을 출력한다.\
front: 덱의 가장 앞에 있는 정수를 출력한다. 만약 덱에 들어있는 정수가 없는 경우에는 -1을 출력한다.\
back: 덱의 가장 뒤에 있는 정수를 출력한다. 만약 덱에 들어있는 정수가 없는 경우에는 -1을 출력한다.
- 입력\
첫째 줄에 주어지는 명령의 수 N (1 ≤ N ≤ 10,000)이 주어진다. 둘째 줄부터 N개의 줄에는 명령이 하나씩 주어진다. 주어지는 정수는 1보다 크거나 같고, 100,000보다 작거나 같다. 문제에 나와있지 않은 명령이 주어지는 경우는 없다.

- 출력\
출력해야하는 명령이 주어질 때마다, 한 줄에 하나씩 출력한다.

In [31]:
from collections import deque
n = int(input())
lst = [list(input().split()) for _ in range(n)]

In [49]:
lst

[['push_back', '1'],
 ['push_front', '2'],
 ['front'],
 ['back'],
 ['size'],
 ['empty'],
 ['pop_front'],
 ['pop_back'],
 ['pop_front'],
 ['size'],
 ['empty'],
 ['pop_back'],
 ['push_front', '3'],
 ['empty'],
 ['front']]

In [54]:
from collections import deque
deque = deque()
for i in lst:
    if i[0] == 'push_front':
        deque.appendleft(int(i[1]))
    elif i[0] == 'push_back':
        deque.append(int(i[1]))
    elif i[0] == 'pop_front':
        if len(deque) == 0:
            print(-1)
        else:
            a = deque.popleft()
            print(a)
    elif i[0] == 'pop_back':
        if len(deque) == 0:
            print(-1)
        else:
            a = deque.pop()
            print(a)
    elif i[0] == 'size':
        print(len(deque))
    elif i[0] == 'empty':
        if len(deque) == 0:
            print(1)
        else:
            print(0)
    elif i[0] == 'front':
        if len(deque) == 0:
            print(-1)
        else:
            print(deque[0])
    elif i[0] == 'back':
        if len(deque) == 0:
            print(-1)
        else:
            print(deque[len(deque)-1])


2
1
2
0
2
1
-1
0
1
-1
0
3


## 통계학
- 문제\
수를 처리하는 것은 통계학에서 상당히 중요한 일이다. 통계학에서 N개의 수를 대표하는 기본 통계값에는 다음과 같은 것들이 있다. 단, N은 홀수라고 가정하자.\
산술평균 : N개의 수들의 합을 N으로 나눈 값\
중앙값 : N개의 수들을 증가하는 순서로 나열했을 경우 그 중앙에 위치하는 값\
최빈값 : N개의 수들 중 가장 많이 나타나는 값\
범위 : N개의 수들 중 최댓값과 최솟값의 차이\
N개의 수가 주어졌을 때, 네 가지 기본 통계값을 구하는 프로그램을 작성하시오.

- 입력\
첫째 줄에 수의 개수 N(1 ≤ N ≤ 500,000)이 주어진다. 단, N은 홀수이다. 그 다음 N개의 줄에는 정수들이 주어진다. 입력되는 정수의 절댓값은 4,000을 넘지 않는다.

- 출력\
첫째 줄에는 산술평균을 출력한다. 소수점 이하 첫째 자리에서 반올림한 값을 출력한다.
둘째 줄에는 중앙값을 출력한다.
셋째 줄에는 최빈값을 출력한다. 여러 개 있을 때에는 최빈값 중 두 번째로 작은 값을 출력한다.
넷째 줄에는 범위를 출력한다.

In [44]:
n = int(input())
lst = list(int(input()) for _ in range(n))

In [41]:
# 산술평균
print(round(sum(lst)/n)) 

# 중앙값
lst.sort()
# 홀수개면 (n+1)/2 번째가 중앙값
med_id = int((n+1)/2 - 1)
print(lst[med_id])

# 최빈값
count = {}
lst1 = list(set(lst))
for i in lst1:
    count[i] = lst.count(i)
count = sorted(count.items(), key=lambda x: (-x[1], x[0]))
if len(count) == 1:
    print(count[0][0])
elif count[0][0] == count[1][0]:
    print(count[1][0])# 여러개 있을 시, 최빈값 중 두 번째로 작은 값
else:
    print(count[0][0]) 

# 범위
print(lst[n-1]-lst[0]) 

## 시간초과; 아마 최빈값에서 시간초과 난 듯..? (for문, sort)

0
0
0
1


In [45]:
from collections import Counter
# 산술평균
print(round(sum(lst)/n)) 

# 중앙값
lst.sort()
# 홀수개면 (n+1)/2 번째가 중앙값
med_id = int((n+1)/2 - 1)
print(lst[med_id])

# 최빈값
mode_value = Counter(lst).most_common() # Counter 내장함수: 값에 개수를 연결시켜주며 최빈값 기준 정렬 후, 같으면 값이 작은 것부터 정렬됨
if len(mode_value)>1 and mode_value[0][1] == mode_value[1][1]: # 최빈값이 2개 이상이면
    print(mode_value[1][0])
else:
    print(mode_value[0][0])


# 범위
print(lst[n-1]-lst[0]) 

## 내장함수 사용하여 시간초과 문제 해결

-2
-2
-1
2


## 요세푸스 문제
- 문제\
요세푸스 문제는 다음과 같다.\
1번부터 N번까지 N명의 사람이 원을 이루면서 앉아있고, 양의 정수 K(≤ N)가 주어진다. 이제 순서대로 K번째 사람을 제거한다. 한 사람이 제거되면 남은 사람들로 이루어진 원을 따라 이 과정을 계속해 나간다. 이 과정은 N명의 사람이 모두 제거될 때까지 계속된다. 원에서 사람들이 제거되는 순서를 (N, K)-요세푸스 순열이라고 한다. 예를 들어 (7, 3)-요세푸스 순열은 <3, 6, 2, 7, 5, 1, 4>이다.\
N과 K가 주어지면 (N, K)-요세푸스 순열을 구하는 프로그램을 작성하시오.

- 입력\
첫째 줄에 N과 K가 빈 칸을 사이에 두고 순서대로 주어진다. (1 ≤ K ≤ N ≤ 5,000)

- 출력\
예제와 같이 요세푸스 순열을 출력한다.

In [102]:
n, k = map(int, input().split())

# 1부터 n까지의 리스트와 제거된 사람의 리스트를 만듦
lst = list(range(1,n+1))
pop_lst = []

id = -1
while len(pop_lst) != n:
    move = 0
    while move != k:
        move += 1
        id += 1 # 인덱스 한칸 뒤로
        if id > n-1: # id가 n-1(lst의 최대인덱스) 넘어가면 다시 처음으로 돌아가 넘어간 만큼 가야함
            id -= n

        while lst[id] in pop_lst: # 만약 해당 인덱스가 이미 제거되었다면 한칸 더 뒤로
            id += 1
            if id > n-1: # id가 n-1(lst의 최대인덱스) 넘어가면 다시 처음으로 돌아가 넘어간 만큼 가야함
                id -= n
    
    pop_lst.append(lst[id])

for i in range(len(pop_lst)):
    if i == 0:
        print('<', pop_lst[i], ',', sep='', end=' ')
    elif i == n-1:
        print(pop_lst[i], '>', sep='', end=' ')
    else:
        print(pop_lst[i], ',', sep='', end=' ')
        
# 시간초과; 환형 큐 이용해서 풀어보기

[3, 6, 2, 7, 5, 1, 4]


## 방 번호
- 문제\
다솜이는 은진이의 옆집에 새로 이사왔다. 다솜이는 자기 방 번호를 예쁜 플라스틱 숫자로 문에 붙이려고 한다.\
다솜이의 옆집에서는 플라스틱 숫자를 한 세트로 판다. 한 세트에는 0번부터 9번까지 숫자가 하나씩 들어있다. 다솜이의 방 번호가 주어졌을 때, 필요한 세트의 개수의 최솟값을 출력하시오. (6은 9를 뒤집어서 이용할 수 있고, 9는 6을 뒤집어서 이용할 수 있다.)

- 입력\
첫째 줄에 다솜이의 방 번호 N이 주어진다. N은 1,000,000보다 작거나 같은 자연수이다.

- 출력\
첫째 줄에 필요한 세트의 개수를 출력한다.

In [15]:
from collections import Counter
import math
num = list(input())

# 6과 9는 같은 카드로 보면 되므로 모두 6으로 통일. 즉, 9가 나오면 6으로 바꿔주기
for i in range(len(num)):
    if num[i] == '9':
        num[i] = '6'
# 카드들 중 최빈값을 찾아 그 개수만큼 세트를 사면 됨. 단, 6이 최빈값이라면 개수/2를 올림한 만큼 사야되므로 두번째 최빈값가 비교하여 더 큰 개수만큼 사기
mode_value = Counter(num).most_common()
if mode_value[0][0] != '6': # 최빈값인 카드가 6이 아니면 그대로 개수 출력
    print(mode_value[0][1])
else: # 최빈값인 카드가 6이라면 개수/2를 올림한 수와 두번째 최빈값의 개수를 비교해서 더 큰 값을 출력
    if int(math.ceil(mode_value[0][1]/2)) >= mode_value[1][1]:
        print(int(math.ceil(mode_value[0][1]/2)))
    else:
        print(mode_value[1][1])



2


## 프린터 큐
- 문제\
여러분도 알다시피 여러분의 프린터 기기는 여러분이 인쇄하고자 하는 문서를 인쇄 명령을 받은 ‘순서대로’, 즉 먼저 요청된 것을 먼저 인쇄한다. 여러 개의 문서가 쌓인다면 Queue 자료구조에 쌓여서 FIFO - First In First Out - 에 따라 인쇄가 되게 된다. 하지만 상근이는 새로운 프린터기 내부 소프트웨어를 개발하였는데, 이 프린터기는 다음과 같은 조건에 따라 인쇄를 하게 된다.\
현재 Queue의 가장 앞에 있는 문서의 ‘중요도’를 확인한다.\
나머지 문서들 중 현재 문서보다 중요도가 높은 문서가 하나라도 있다면, 이 문서를 인쇄하지 않고 Queue의 가장 뒤에 재배치 한다. 그렇지 않다면 바로 인쇄를 한다.\
예를 들어 Queue에 4개의 문서(A B C D)가 있고, 중요도가 2 1 4 3 라면 C를 인쇄하고, 다음으로 D를 인쇄하고 A, B를 인쇄하게 된다.\
여러분이 할 일은, 현재 Queue에 있는 문서의 수와 중요도가 주어졌을 때, 어떤 한 문서가 몇 번째로 인쇄되는지 알아내는 것이다. 예를 들어 위의 예에서 C문서는 1번째로, A문서는 3번째로 인쇄되게 된다.\

- 입력\
첫 줄에 테스트케이스의 수가 주어진다. 각 테스트케이스는 두 줄로 이루어져 있다.\
테스트케이스의 첫 번째 줄에는 문서의 개수 N(1 ≤ N ≤ 100)과, 몇 번째로 인쇄되었는지 궁금한 문서가 현재 Queue에서 몇 번째에 놓여 있는지를 나타내는 정수 M(0 ≤ M < N)이 주어진다. 이때 맨 왼쪽은 0번째라고 하자. 두 번째 줄에는 N개 문서의 중요도가 차례대로 주어진다. 중요도는 1 이상 9 이하의 정수이고, 중요도가 같은 문서가 여러 개 있을 수도 있다.

- 출력\
각 테스트 케이스에 대해 문서가 몇 번째로 인쇄되는지 출력한다.



In [14]:
from collections import deque

n = int(input())
for i in range(n):
    n_doc, curious = map(int, input().split())
    lst = list(map(int, input().split()))
    deq = deque(lst)
    count = 0 # 인쇄된 횟수
    while True:
        if len(deq) > 0 and deq[0] != max(deq): # 큐의 첫번째 원소가 최대값이 아니라면 왼쪽에서 빼서 오른쪽으로 붙이기
            deq.append(deq.popleft())
            curious -= 1 # 궁금한 문서의 인덱스도 같이 옮겨야 함
            if curious < 0: # 만일 인덱스가 음수가 된다면 맨 끝 인덱스로 옮겨야 함
                curious += len(deq)
        elif len(deq) > 0 and deq[0] == max(deq): # 최대값이 맞다면 인쇄, 즉 popleft
            if curious == 0: # 궁금한 인덱스가 0번째면 인쇄할 문서가 궁금한 문서라는 것이므로 count에 1을 더해 출력하고 반복문 마침
                count += 1
                print(count)
                break
            else: # 궁금한 인덱스가 0이 아니라면 그냥 인쇄. 즉 popleft하고 인덱스 하나씩 밀리므로 궁금한 인덱스도 -1
                deq.popleft()
                count += 1
                curious -= 1
        if len(deq) == 0:
            break


1
2
5
